File: app/src/main/java/com/example/dagger_hilt/domain/NetworkBoundResource.java
package com.example.dagger_hilt.domain;

import androidx.annotation.MainThread;
import androidx.annotation.WorkerThread;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;

import com.example.dagger_hilt.data.datasource.web.models.response.ApiEmptyResponse;
import com.example.dagger_hilt.data.datasource.web.models.response.ApiErrorResponse;
import com.example.dagger_hilt.data.datasource.web.models.response.ApiResponse;
import com.example.dagger_hilt.data.datasource.web.models.response.ApiSuccessResponse;
import com.example.dagger_hilt.sys.util.AppExecutors;
import com.example.dagger_hilt.sys.util.Resource;

public abstract class NetworkBoundResource<ResultType, RequestType> {
    private final AppExecutors appExecutors;
    private final MediatorLiveData<Resource<ResultType>> result = new MediatorLiveData<>();

    public NetworkBoundResource(AppExecutors appExecutors) {
        this.appExecutors = appExecutors;
        setValue(Resource.loading(null));
        var dbSource = loadFromDb();
        result.addSource(dbSource, (data) -> {
            result.removeSource(dbSource);
            if (shouldFetch(data)) {
                fetchFromNetwork(dbSource);
            } else {
                result.addSource(dbSource, (newData) -> setValue(Resource.success(newData)));
            }
        });
    }

    private void fetchFromNetwork(LiveData<ResultType> dbSource) {
        LiveData<ApiResponse> apiResponse;
        apiResponse = createCall();

        result.addSource(dbSource, (newData -> setValue(Resource.loading(newData))));
        result.addSource(apiResponse, response -> {
            result.removeSource(apiResponse);
            result.removeSource(dbSource);
            if (response instanceof ApiSuccessResponse<?>)
                appExecutors.getDiskIO().execute(() -> {
                            saveCallResult(processResponse((ApiSuccessResponse<RequestType>) response));
                            appExecutors.getMainThread().execute(() ->
                                    result.addSource(loadFromDb(), (newData) ->
                                            setValue(Resource.success(newData))));
                        }
                );
            else if (response instanceof ApiEmptyResponse)
                appExecutors.getMainThread().execute(() ->
                        result.addSource(loadFromDb(), (newData) ->
                                setValue(Resource.success(newData))));
            else if (response instanceof ApiErrorResponse) {
                onFetchFailed();
                result.addSource(dbSource, (newData ->
                        setValue(Resource.error(((ApiErrorResponse) response)
                                .getErrorMessage(), newData))));
            }
        });
    }

    public LiveData<Resource<ResultType>> asLiveData() {
        return result;
    }

    @MainThread
    private void setValue(Resource<ResultType> newValue) {
        if (result.getValue() != newValue) {
            result.setValue(newValue);
        }
    }

    protected void onFetchFailed() {

    }

    @WorkerThread
    protected abstract void saveCallResult(RequestType requestType);

    @WorkerThread
    protected RequestType processResponse(ApiSuccessResponse<RequestType> response) {
        return response.getBody();
    }

    @MainThread
    protected abstract LiveData<ApiResponse> createCall();

    @MainThread
    protected abstract boolean shouldFetch(ResultType data);

    @MainThread
    protected abstract LiveData<ResultType> loadFromDb();

}

File: app/src/main/java/com/example/gob_fact/domain/NetworkBoundResource.kt
package com.example.gob_fact.domain

import androidx.annotation.MainThread
import androidx.annotation.WorkerThread
import androidx.lifecycle.LiveData
import androidx.lifecycle.MediatorLiveData
import com.example.gob_fact.data.datasource.web.models.response.ApiEmptyResponse
import com.example.gob_fact.data.datasource.web.models.response.ApiErrorResponse
import com.example.gob_fact.data.datasource.web.models.response.ApiResponse
import com.example.gob_fact.data.datasource.web.models.response.ApiSuccessResponse
import com.example.gob_fact.sys.util.AppExecutors
import com.example.gob_fact.sys.util.Resource as Resource

abstract class NetworkBoundResource<ResultType, RequestType> @MainThread constructor(
    private val appExecutors: AppExecutors
) {

    private val result = MediatorLiveData<Resource<ResultType>>()

    init {
        @Suppress("LeakingThis")
        val dbSource = loadFromDb()
        result.addSource(dbSource) { data ->
            result.value = Resource.loading(null)
            result.removeSource(dbSource)
            if (shouldFetch(data))
                fetchFromNetwork(dbSource)
            else result.addSource(dbSource) { newData ->
                setValue(Resource.success(newData))
            }

        }
    }

    @MainThread
    private fun setValue(newValue: Resource<ResultType>) {
        if (result.value != newValue) {
            result.value = newValue
        }
    }

    private fun fetchFromNetwork(dbSource: LiveData<ResultType>) {
        val apiResponse: LiveData<ApiResponse<RequestType>> = createCall()
        result.addSource(dbSource) { newData ->
            setValue(Resource.loading(newData))
        }
        result.addSource(apiResponse) { response ->
            result.removeSource(apiResponse)
            result.removeSource(dbSource)
            when (response) {
                is ApiSuccessResponse -> {
                    appExecutors.diskIO().execute {
                        saveCallResult(processResponse(response))
                        appExecutors.mainThread().execute {
                            result.addSource(loadFromDb()) { newData ->
                                setValue(Resource.success(newData))
                            }
                        }
                    }
                }

                is ApiEmptyResponse -> {
                    appExecutors.mainThread().execute {
                        // reload from disk whatever we had
                        result.addSource(loadFromDb()) { newData ->
                            setValue(Resource.success(newData))
                        }
                    }
                }

                is ApiErrorResponse -> {
                    onFetchFailed()
                    result.addSource(dbSource) { newData ->
                        setValue(Resource.error(response.errorMessage, newData))
                    }
                }
            }
        }
    }

    protected open fun onFetchFailed() {}

    fun asLiveData() = result as LiveData<Resource<ResultType>>

    @WorkerThread
    protected open fun processResponse(response: ApiSuccessResponse<RequestType>): RequestType =
        response.body

    @WorkerThread
    protected abstract fun saveCallResult(response: RequestType)

    @MainThread
    protected abstract fun shouldFetch(data: ResultType?): Boolean

    @MainThread
    protected abstract fun loadFromDb(): LiveData<ResultType>

    @MainThread
    protected abstract fun createCall(): LiveData<ApiResponse<RequestType>>
}
